internal/osbuild/assembler_test.go

```diff
@@ -0,0 +1,230 @@
+			},
+		},
+		{
+			name: "rawfs assembler full",
+			assembler: Assembler{
+				Name: "org.osbuild.rawfs",
+				Options: &RawFSAssemblerOptions{
+					Filename:           "filesystem.img",
+					RootFilesystemUUID: uuid.MustParse("76a22bf4-f153-4541-b6c7-0332c0dfaeac"),
+					Size:               2147483648,
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.rawfs","options":{"filename":"filesystem.img","root_fs_uuid":"76a22bf4-f153-4541-b6c7-0332c0dfaeac","size":2147483648}}`),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			assembler := &tt.assembler
+			var gotAssembler Assembler
+			err := gotAssembler.UnmarshalJSON(tt.args.data)
+			assert.Equalf((err != nil), tt.wantErr, "Assembler.UnmarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
+			if tt.wantErr {
+				return
+			}
+			gotBytes, err := json.Marshal(assembler)
+			assert.Nilf(err, "Could not marshal assembler: %v", err)
+			assert.Equalf(gotBytes, tt.args.data, "Expected `%v`, got `%v`", string(tt.args.data), string(gotBytes))
+			assert.Truef(reflect.DeepEqual(&gotAssembler, assembler),
+				"got {%v, %v}, expected {%v, %v}",
+				gotAssembler.Name,
+				gotAssembler.Options,
+				assembler.Name,
+				assembler.Options,
+			)
+		})
+	}
+}
+
+func TestNewQEMUAssembler(t *testing.T) {
+	options := &QEMUAssemblerOptions{
+		Format:   "qcow2",
+		Filename: "disk.qcow2",
+		Size:     2147483648,
+		PTUUID:   "0x14fc63d2",
+		PTType:   "mbr",
+		Partitions: []QEMUPartition{QEMUPartition{
+			Start:    2048,
+			Bootable: true,
+			Filesystem: QEMUFilesystem{
+				Type:       "ext4",
+				UUID:       "76a22bf4-f153-4541-b6c7-0332c0dfaeac",
+				Label:      "root",
+				Mountpoint: "/",
+			},
+		}},
+	}
+	expectedAssembler := &Assembler{
+		Name: "org.osbuild.qemu",
+		Options: &QEMUAssemblerOptions{
+			Format:   "qcow2",
+			Filename: "disk.qcow2",
+			Size:     2147483648,
+			PTUUID:   "0x14fc63d2",
+			PTType:   "mbr",
+			Partitions: []QEMUPartition{QEMUPartition{
+				Start:    2048,
+				Bootable: true,
+				Filesystem: QEMUFilesystem{
+					Type:       "ext4",
+					UUID:       "76a22bf4-f153-4541-b6c7-0332c0dfaeac",
+					Label:      "root",
+					Mountpoint: "/",
+				},
+			}},
+		},
+	}
+	assert.Equal(t, expectedAssembler, NewQEMUAssembler(options))
+}
+
+func TestNewTarAssembler(t *testing.T) {
+	options := NewTarAssemblerOptions("disk.tar.xz", 0)
+	options.SetCompression("xz")
+	expectedAssembler := &Assembler{
+		Name: "org.osbuild.tar",
+		Options: &TarAssemblerOptions{
+			Filename:    "disk.tar.xz",
+			Size:        0,
+			Compression: "xz",
+		},
+	}
+	assert.Equal(t, expectedAssembler, NewTarAssembler(options))
+}
+
+func TestNewRawFSAssembler(t *testing.T) {
+	options := &RawFSAssemblerOptions{
+		Filename:           "disk.img",
+		RootFilesystemUUID: uuid.MustParse("76a22bf4-f153-4541-b6c7-0332c0dfaeac"),
+		Size:               0,
+		FilesystemType:     "ext4",
+	}
```

Comment:
Like mentioned in my previous comment: I think this is not necessary for the current scenario(s). The constructor is passing the options argument unmodified to the Assembler class so it actually should be possible to exercise and validate the functionality of all `New` functions without any kind of options (empty or nil struct even).